### PR TITLE
[CodeGen] Rework MVT representation of capabilities and add type inference

### PR DESCRIPTION
**Pull request by jrtc27**
This more closely aligns how capabilites are handled with iPTR and iN
types, introducing a new cPTR wildcard and fixed-width cN types that it
maps to, with iPTRAny being iPTR and cPTR. As a result we no longer need
to duplicate patterns for each fixed-width type in certain cases, and
the few upstream MIPS tests we've long marked as XFAIL'ed now pass.

As part of this, a few of the APIs around MVTs have been renamed from
fat pointer terminology to use capability instead. Where relatively
straightforward to do these have then been adopted throughout, but some
are pervasive and have deprecated aliases under the old names, whose
uses should be incrementally reduced.

### CODE REVIEW

**Comment by jrtc27** [comment]
NB: I have not yet tested CheriBSD with this.

File: llvm/utils/TableGen/CodeGenDAGPatterns.cpp [hunk 1]
```diff
@@ -271,20 +271,72 @@ void TypeSetByHwMode::dump() const {
   dbgs() << *this << '\n';
 }
 
-bool TypeSetByHwMode::intersect(SetType &Out, const SetType &In) {
-  bool OutP = Out.count(MVT::iPTR), InP = In.count(MVT::iPTR);
-  // Complement of In.
-  auto CompIn = [&In](MVT T) -> bool { return !In.count(T); };
+namespace {
+enum class TypeCat {
```
Comment (jrtc27) [inline llvm/utils/TableGen/CodeGenDAGPatterns.cpp:275]: I will likely rewrite this to be less overkill with the templates. At one point I thought it was going to be a lot more complicated, but as it turns out the function really just needs an MVT and an MVT->bool.

File: llvm/utils/TableGen/CodeGenDAGPatterns.cpp [hunk 2]
```diff
@@ -1660,14 +1703,11 @@ bool SDTypeConstraint::ApplyTypeConstraint(TreePatternNode *N,
...
     // If we support fat pointers, then this narrows it down to two types.
-    if (TP.getDAGPatterns().enableFatPointers()) {
-      // FIXME: We should be able to do the type inference correctly from the
-      // two types, but for now just disable this constraint on architectures
-      // with fat pointers.
-      return false;
-      ArrayRef<ValueTypeByHwMode> PtrTys({ValueTypeByHwMode(MVT::iFATPTRAny),
-                                          ValueTypeByHwMode(MVT::iPTR)});
-      return NodeToApply->UpdateNodeType(ResNo, PtrTys, TP);
+    // TODO: Make this less ugly
```
Comment (jrtc27) [inline llvm/utils/TableGen/CodeGenDAGPatterns.cpp:1706]: I intend to fix this before it gets merged

**Review by arichardson — APPROVED** [review]
This looks like a huge improvement. I have never looked at those tablegen internals, so if it works that sounds good to me.

One or two of the clang format suggestions are legitimate since the shorter name now makes the line fit but most of them are quite odd suggestions...

File: llvm/lib/Target/Mips/MipsInstrCheri.td [hunk 3]
```diff
@@ -939,88 +935,76 @@ def CSCB  : CheriFmtCSC<0, (outs GPR32Opnd:$rd), (ins GPR32Opnd: $rs, CheriOpnd0
...
+defm: FrameIndexLoadPatterns<zextloadi1, i64, CAPLOADU8, simm8Pattern>;
+defm: FrameIndexLoadPatterns<sextloadi1, i32, CAPLOAD832, simm8Pattern>;
+defm: FrameIndexLoadPatterns<sextloadi1, i64, CAPLOAD8, simm8Pattern>;
+// Use clbu instead of clb for anyext to match MIPS
+defm: FrameIndexLoadPatterns<extloadi1, i32, CAPLOADU832, simm8Pattern>;
+defm: FrameIndexLoadPatterns<extloadi1, i64, CAPLOADU8, simm8Pattern>;
+
+// stores
+
+// This pattern allows omitting lots of unnncessary sll instructions
```
[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Target/Mips/MipsInstrCheri.td:996]: Eh, I wanted to minimally touch CHERI-MIPS beyond adopting the new approach... ditto your comment below about deleting commented code

[the remark this answers is not in the document]
Comment (arichardson) [reply llvm/lib/Target/Mips/MipsInstrCheri.td:996]: Fair, let's just delete it all at once :)

**Comment by jrtc27** [comment]
Clean builds of cheribsd-riscv64-purecap have cheribsdtest-* all passing on CHERI-QEMU and CHERI-PURECAP-QEMU kernels.

**Comment by jrtc27** [comment]
FYI https://github.com/llvm/llvm-project/pull/81688 for upstreaming the intersect redesign.

**Pull request merged**